### PR TITLE
[JUJU-3749] Add SupportedBases to ResolveCharms output

### PR DESCRIPTION
**Pull request by jack-w-shaw**
This is a step towards replacing series with bases in Juju

One of the returns of ResolveCharms was a slice of supported series. Add to this Supported bases and make use of this in the client.

As a flyby rev juju/charm to v11 and juju/collections to v1.0.3. For this reason, you may wish to review the PR one commit at a time. Also replace our usage of go-spew with pretty.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- ~[ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- ~[ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

### Verify you can deploy a charm and refresh it

```
juju deploy postgresql --revision 287 --channel stable
juju refresh postgresql --revision 288
```

Repeat this with:
- build from branch client, build from branch controller
- build from branch client, 3.3 branch controller
- 3.3 branch client, build from branch controller

### Verify kubeflow can be deployed

```
juju deploy kubeflow --trust
```
Repeat this with clients both using charms facade version 6 and 7

### CODE REVIEW

**Review by hmlanigan — CHANGES_REQUESTED** [review]
The charms facade should have a new revision for an old and new version of ResolveCharms with different results. It'll be easier to keep client compatibility and remove the old version in juju 4.x

The commit message of 35fdd05, is missing note of the addition of `--force-base` flag to refresh. That could be a separate commit. The QA steps are missing a test for this change.

**Comment by hmlanigan** [comment]
@jack-w-shaw JUJU-3479 is an o'Reilly training Jira ticket, which Jira ticket is this change for? I wanted to add some review time.

**Comment by jack-w-shaw** [comment]
> @jack-w-shaw JUJU-3479 is an o'Reilly training Jira ticket, which Jira ticket is this change for? I wanted to add some review time.

Oops, it's JUJU-3749

File: cmd/juju/application/refresh.go
Comment (jack-w-shaw) [reply cmd/juju/application/refresh.go:384]: Good spot!

**Review by hmlanigan — COMMENTED** [review]
This is what I have so far, still reviewing. It's a large PR. :-)
QA:

`$ juju deploy ubuntu --base ubuntu@14.04
ERROR ubuntu is not available on the following series: trusty not supported
`

The above output comes from composeSuggestions() in core/charm/repository as part of ResolveCharms

File: api/client/charms/client_test.go
Comment (jack-w-shaw) [reply api/client/charms/client_test.go:49]: Addressed, but have left unresolved so you remember to check since I'm not entirely confident with my solution

**Review by hmlanigan — COMMENTED** [review]
Current pieces look good. Lots of changes for the better.

Still the one QA fail needs fixing before approval.

File: go.mod
Comment (jack-w-shaw) [reply go.mod]: Funny this appeared, spew was already a dependency, which was used in one place in one test. I have replaced that with pretty, which moved go-spew back to an indirect

**Comment by jack-w-shaw** [comment]
> QA:
> 
> `$ juju deploy ubuntu --base ubuntu@14.04 ERROR ubuntu is not available on the following series: trusty not supported `
> 
> The above output comes from composeSuggestions() in core/charm/repository as part of ResolveCharms

@hmlanigan  I'm not convinced this is within the scope of this PR.

This error actually comes from a SeriesSelector in `compatibiltiyPrepareAndDeploy`, not anything to do with ResolveCharm. This should be changed to a base selector, but that's the job for another PR imo

**Comment by hmlanigan** [comment]
@jack-w-shaw I got the wrong error. I'm thinking of the code in composeSuggestions() in the core charm repository code. As agreed, there is a new Jira ticket to cover that change https://warthogs.atlassian.net/browse/JUJU-4005

**Review by hmlanigan — APPROVED** [review]
lgtm

**Pull request merged**
